File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.cpp
class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        sort(events.begin(),events.end());
        int n=events.size(),ans=0;
        vector<int> mx(n),st(n);
        mx[n-1]=events[n-1][2];
        for(int i=n-2;i>=0;i--) {
            mx[i]=max(mx[i+1],events[i][2]);
        }
        mx.push_back(0);
        for(int i=0;i<n;i++) {
            st[i]=events[i][0];
        }
        for(int i=0;i<n;i++) {
            int x=upper_bound(st.begin(),st.end(),events[i][1])-st.begin();
            ans=max(ans,events[i][2]+mx[x]);
        }
        return ans;
    }
};

File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.py
class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        events.sort()
        heap = []
        res, curr = 0, 0

        for start, end, value in events:
            while heap and heap[0][0] < start:
                    curr = max(curr, heapq.heappop(heap)[1])
            
            res = max(res, value + curr)
            heapq.heappush(heap, (end, value))
        
        return res
 